[API] Make `Parameterized` refine `Differentiable`

rxwei opened the pull request:
Make `Parameterized` refine `Differentiable` so that values of conforming types can work directly with differentiation. The `CotangentVector` associated type is set to be the nested `Parameters` struct. The synthesized `Parameters` struct must derives conformances to `Differentiable`, `VectorNumeric` (if possible), `KeyPathIterable`, and `ParameterGroup` (if possible). 

For types that conform to `Parameterized`, derived conformances sets the following:
```
@_fieldwiseProductSpace
typealias TangentVector = Parameters
@_fieldwiseProductSpace
typealias CotangentVector = Parameters
```

It should really set `TangentVector` to be `Parameters.TangentVector`, but that change will be made in a separate patch.

This enables differentiation of member accesses, as tested in `test/TensorFlowRuntime/parameterized.swift`. A next step is synthesizing `@differentiable` for all stored properties of `Differentiable` types.

-----

@dan-zheng took over and completed most of the work in this patch. Here are his notes:

`Parameterized` + `Differentiable` derived conformances are still fragile.
`Differentiable` synthesis depends on `Parameterized` synthesis (`Parameters`
struct must exist so that `TangentVector` and `CotangentVector` can be set to
it). This will be handled more reliably in a follow-up.

`Differentiable` derived conformances are also a bit fragile and will be
improved. For now, incorrectly synthesized requirements are manually specified
in tests.

rxwei committed:
WIP: Make `Parameterized` refine `Differentiable`

dan-zheng commented on lib/Sema/DerivedConformanceParameterized.cpp:
```diff
@@ -400,9 +431,34 @@ static Type deriveParameterized_Parameters(DerivedConformance &derived) {
+    return deriveParameterized_VectorSpace(TC.Context.Id_TangentVector, *this);
+  if (requirement->getBaseName() == TC.Context.Id_CotangentVector)
+    return deriveParameterized_VectorSpace(TC.Context.Id_CotangentVector,
+                                           *this);
```
Don't worry about synthesizing requirements defined on `Differentiable`: those will fall out of `Differentiable` derived conformances. The newly added lines here won't trigger.

dan-zheng commented on lib/Sema/DerivedConformances.cpp:
```diff
@@ -330,6 +329,13 @@ ValueDecl *DerivedConformance::getDerivableRequirement(TypeChecker &tc,
+    if (name.isSimpleName(ctx.Id_TangentVector) ||
+        name.isSimpleName(ctx.Id_CotangentVector))
+      // TOOD: Also `getRequirement(KnownProtocolKind::Differentiable)`?
+      return getRequirement(KnownProtocolKind::Parameterized);
```
Similarly, you can remove this code. This will fall out of `Differentiable` derived conformances.

dan-zheng commented on lib/Sema/DerivedConformanceParameterized.cpp:
```diff
@@ -374,12 +382,35 @@ static Type deriveParameterized_Parameters(DerivedConformance &derived) {
-    parametersDecl->setInherited(C.AllocateCopy(inherited));
+  // Declare conformance to `VectorNumeric` (if possible), `Differentiable`
+  // and `KeyPathIterable` protocols. Conformances will be derived.
+  SmallVector<TypeLoc, 3> inherited {differentiableType, keyPathIterableType};
```
This is correct and crucial!
Inheriting `keyPathIterableType` is important to leverage new `KeyPathIterable` protocol.
Inheriting `differentiableType ` is important but it blocked by `Differentiable` derived conformances. I'm working on it and will push ASAP.

dan-zheng commented on stdlib/public/TensorFlow/Parameterized.swift:
```diff
@@ -71,18 +72,18 @@ public extension Parameterized where Parameters : ParameterGroup {
+//      self = newValue
+//    }
+//  }
+//}
```
I think removing this makes sense. It's fine for `Tensor` to not conform to `Parameterized` and to act as a "base case" member.

dan-zheng committed:
[API] Start derived conformances for `Parameterized` + `Differentiable`.

`Parameterized` now refines `Differentiable`.
The synthesized `Parameters` struct must derives conformances to
`Differentiable`, `VectorNumeric`, `KeyPathIterable`, and (if possible)
`ParameterGroup`.

For types that conform to `Parameterized`, derived conformances
sets the following:
```
@_fieldwiseProductSpace
typealias TangentVector = Parameters
@_fieldwiseProductSpace
typealias CotangentVector = Parameters
```

This enables differentiation of member accesses, as tested in
`test/TensorFlowRuntime/parameterized.swift`. A next step is synthesizing
`@differentiable` for all stored properties of `Differentiable` types.

--

`Parameterized` + `Differentiable` derived conformances are still fragile.
`Differentiable` synthesis depends on `Parameterized` synthesis (`Parameters`
struct must exist so that `TangentVector` and `CotangentVector` can be set to
it). This will be handled more reliably in a follow-up.

`Differentiable` derved conformances are also a bit fragile and will be
improved. For now, incorrectly synthesized requirements are manually specified
in tests.

---

This commit also contains many minor but important fixes. Details ommited.

dan-zheng committed:
Merge branch 'tensorflow' into parameterized-refine-differentiable

dan-zheng committed:
Fix `Parameterized` sema tests.

Mark deprecated derived conformances behavior in test file.
They will be handled in a follow-up patch.

dan-zheng commented:
@swift-ci Please test tensorflow

jekbradbury commented:
> Note: Using `Tensor`-typed parameters is blocked by conforming synthesized `Parameters` to `TensorArrayProtocol` for GPE.

Are there any blockers you're aware of in the case of dynamic compilation/no GPE? I'm getting a runtime segfault on this branch with `-tf-dynamic-compilation` for the following code (crash in `__tf_to_host`); will look into it:
```swift
struct Model: Parameterized, Differentiable {
  @TFParameter var w: Tensor<Float>
  func tangentVector(from cotangent: Parameters) -> Parameters {
    return cotangent
  }
}

let model = Model(w: Tensor<Float>(zeros: [5]))
let pb = pullback(at: model) { m in m.w }
print(pb(Tensor<Float>(zeros: [5])))
```
(although this works:)
```swift
let model = Model(w: Tensor<Float>(zeros: [5]))
let x = Tensor<Float>(zeros: [5])
let pb = pullback(at: model) { m in m.w + x }
print(pb(Tensor<Float>(zeros: [5])))
```

jekbradbury commented:
Looks like this was because I didn't rebuild the standard library.

rxwei committed:
Clean up

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow
